InstanceContext.Data pointers may become stale and cause panics

### Describe the bug

Using a uintptr to work around CGo FFI limitations in PR #85 seemed like a good idea, and the tests are passing, but when I went to update [Gossamer's runtime](https://github.com/ChainSafe/gossamer) with the latest code I found that this approach has introduced a very serious bug that now affects all InstanceContext Data. I get various unpredictable panics when applying a type assertion on `InstanceContext.Data()`. Among the panics that I observed, I believe the most accurate error message that I got was

```
fatal error: exitsyscall: syscall frame is no longer valid
```

When the Go runtime has to grow a stack, the stack is copied and all pointers on the stack must be updated to the offset from the proper start of the new stack. However `uintptr` is not treated specially in Go, and so is not updated when this occurs. This results in the pointer becoming invalid. 

Some useful further reading about this sort of issue:
https://grokbase.com/t/gg/golang-nuts/147qqaky1h/go-nuts-uintptr-to-go-object-unsafe-to-pass-to-any-function-including-syscalls-c-functions
https://blog.wallaroolabs.com/2018/04/adventures-with-cgo-part-1--the-pointering/
https://blog.cloudflare.com/how-stacks-are-handled-in-go/

### Steps to reproduce

At the moment none of the tests in this repo reproduce the issue and I am working on adding some tests that panic reliably due to this issue.

However it can be easily reproduced in the Gossamer runtime tests when using the latest version of this package. I have a fork with a branch that reproduces this:

```
$ git clone https://github.com/AdamSLevy/gossamer.git
$ cd gossamer
$ git checkout RemoveRuntimeRegistry
$ go test ./runtime/...
t=2019-12-03T13:32:21-0900 lvl=dbug msg=[NewRuntime] runtimeCtx="{trie:0xc0000bd240 allocator:0xc00017e000 keystore:0xc0000bd260}"
t=2019-12-03T13:32:21-0900 lvl=trce msg="[ext_malloc] executing..."
fatal error: exitsyscall: syscall frame is no longer valid

goroutine 20 [running, locked to thread]:
runtime.throw(0x9a4891, 0x2d)
	/usr/lib/go/src/runtime/panic.go:774 +0x72 fp=0xc00018fb50 sp=0xc00018fb20 pc=0x435e02
runtime.exitsyscall()
	/usr/lib/go/src/runtime/proc.go:2957 +0x275 fp=0xc00018fb80 sp=0xc00018fb50 pc=0x43ead5
runtime.cgocall(0x871870, 0xc00018fbe8, 0xc)
	/usr/lib/go/src/runtime/cgocall.go:138 +0x9e fp=0xc00018fbb8 sp=0xc00018fb80 pc=0x4097ce
github.com/wasmerio/go-ext-wasm/wasmer._Cfunc_wasmer_instance_call(0x23e9520, 0x24a4b80, 0xc0001723c0, 0xc000000002, 0xc00016c2f0, 0x1, 0xc000000002)
	_cgo_gotypes.go:730 +0x4d fp=0xc00018fbe8 sp=0xc00018fbb8 pc=0x6421bd
github.com/wasmerio/go-ext-wasm/wasmer.cWasmerInstanceCall(...)
	/home/aslevy/go/pkg/mod/github.com/wasmerio/go-ext-wasm@v0.0.0-20191129141045-bf837c2dd588/wasmer/bridge.go:381
github.com/wasmerio/go-ext-wasm/wasmer.getExportsFromInstance.func1(0xc000180080, 0x2, 0x2, 0x0, 0x0, 0x0, 0x0)
	/home/aslevy/go/pkg/mod/github.com/wasmerio/go-ext-wasm@v0.0.0-20191129141045-bf837c2dd588/wasmer/instance.go:359 +0x35e fp=0xc00018fdb8 sp=0xc00018fbe8 pc=0x64326e
github.com/ChainSafe/gossamer/runtime.(*Runtime).Exec(0xc00016eb40, 0x99187c, 0xc, 0x1, 0xc00018fee8, 0x0, 0x0, 0xf, 0xd3bc3f, 0x22, ...)
	/home/aslevy/repos/gossamer/runtime/wasmer.go:130 +0x1c8 fp=0xc00018fe88 sp=0xc00018fdb8 pc=0x85cee8
github.com/ChainSafe/gossamer/runtime.TestExecVersion(0xc00010a200)
	/home/aslevy/repos/gossamer/runtime/wasmer_test.go:111 +0x1d0 fp=0xc00018ff70 sp=0xc00018fe88 pc=0x85db80
testing.tRunner(0xc00010a200, 0x9b37f0)
	/usr/lib/go/src/testing/testing.go:909 +0xc9 fp=0xc00018ffd0 sp=0xc00018ff70 pc=0x504ef9
runtime.goexit()
	/usr/lib/go/src/runtime/asm_amd64.s:1357 +0x1 fp=0xc00018ffd8 sp=0xc00018ffd0 pc=0x464de1
created by testing.(*T).Run
	/usr/lib/go/src/testing/testing.go:960 +0x350

goroutine 1 [chan receive]:
testing.(*T).Run(0xc00010a200, 0x9933e2, 0xf, 0x9b37f0, 0x489e86)
	/usr/lib/go/src/testing/testing.go:961 +0x377
testing.runTests.func1(0xc00010a100)
	/usr/lib/go/src/testing/testing.go:1202 +0x78
testing.tRunner(0xc00010a100, 0xc000117dc0)
	/usr/lib/go/src/testing/testing.go:909 +0xc9
testing.runTests(0xc0000bd220, 0xd4ba60, 0x17, 0x17, 0x0)
	/usr/lib/go/src/testing/testing.go:1200 +0x2a7
testing.(*M).Run(0xc000106100, 0x0)
	/usr/lib/go/src/testing/testing.go:1117 +0x176
main.main()
	_testmain.go:88 +0x135
FAIL	github.com/ChainSafe/gossamer/runtime	0.348s
ok  	github.com/ChainSafe/gossamer/runtime/allocator	(cached)
FAIL
```

Alternatively you may see this panic instead:
```
t=2019-12-03T13:32:29-0900 lvl=dbug msg=[NewRuntime] runtimeCtx="{trie:0xc00000f280 allocator:0xc000164000 keystore:0xc00000f2a0}"
t=2019-12-03T13:32:29-0900 lvl=trce msg="[ext_malloc] executing..."
--- FAIL: TestExecVersion (0.32s)
panic: <invalid reflect.Value> [recovered]
	panic: <invalid reflect.Value>

goroutine 7 [running]:
testing.tRunner.func1(0xc0000f4200)
	/usr/lib/go/src/testing/testing.go:874 +0x3a3
panic(0x8d3260, 0xc0001680b0)
	/usr/lib/go/src/runtime/panic.go:679 +0x1b2
github.com/ChainSafe/gossamer/runtime.ext_malloc(0x292e4c0, 0xc000000001, 0xc0001759e8)
	/home/aslevy/repos/gossamer/runtime/imports.go:87 +0x222
github.com/ChainSafe/gossamer/runtime._cgoexpwrap_09d97fc801d6_ext_malloc(0x292e4c0, 0x1, 0x0)
	_cgo_gotypes.go:210 +0x33
github.com/wasmerio/go-ext-wasm/wasmer._Cfunc_wasmer_instance_call(0x7ff79801d710, 0x2925d20, 0xc0001583c0, 0xc000000002, 0xc0001522f0, 0x1, 0xc000000000)
	_cgo_gotypes.go:730 +0x4d
github.com/wasmerio/go-ext-wasm/wasmer.cWasmerInstanceCall(...)
	/home/aslevy/go/pkg/mod/github.com/wasmerio/go-ext-wasm@v0.0.0-20191129141045-bf837c2dd588/wasmer/bridge.go:381
github.com/wasmerio/go-ext-wasm/wasmer.getExportsFromInstance.func1(0xc000166080, 0x2, 0x2, 0x0, 0x0, 0x0, 0x0)
	/home/aslevy/go/pkg/mod/github.com/wasmerio/go-ext-wasm@v0.0.0-20191129141045-bf837c2dd588/wasmer/instance.go:359 +0x35e
github.com/ChainSafe/gossamer/runtime.(*Runtime).Exec(0xc000154b40, 0x99187c, 0xc, 0x1, 0xc000175ee8, 0x0, 0x0, 0xf, 0xd3bc3f, 0x22, ...)
	/home/aslevy/repos/gossamer/runtime/wasmer.go:130 +0x1c8
github.com/ChainSafe/gossamer/runtime.TestExecVersion(0xc0000f4200)
	/home/aslevy/repos/gossamer/runtime/wasmer_test.go:111 +0x1d0
testing.tRunner(0xc0000f4200, 0x9b37f0)
	/usr/lib/go/src/testing/testing.go:909 +0xc9
created by testing.(*T).Run
	/usr/lib/go/src/testing/testing.go:960 +0x350
FAIL	github.com/ChainSafe/gossamer/runtime	0.351s
ok  	github.com/ChainSafe/gossamer/runtime/allocator	(cached)
FAIL
```
After re-running the tests about 4 times you will likely see both panics.


### Additional context

[Originally I had proposed using a "registry" map](https://github.com/wasmerio/go-ext-wasm/pull/83) to store InstanceContext data without passing any pointers across the CGo FFI but @losfair [suggested using `uintptr`](https://github.com/wasmerio/go-ext-wasm/pull/83#issuecomment-554975030). This seemed to work until I came across this issue.

I don't think that there is a safe way to pass Go pointers across the CGo boundary, and even if there were a hack that could allow this to work, Go may change the rules around how the CGo FFI works, as they have done in the past. Using a registry map avoids all of those complex issues. 

I am going to open a new PR that implements the registry map instead of the current approach. I highly recommend getting this merged soon because now all InstanceContext.Data may suffer from this issue, not just those data types that the PR was originally supposed to address.

Apologies for not more thoroughly testing the current approach.

--- a/wasmer/import.go
+++ b/wasmer/import.go
@@ -338,10 +338,11 @@ func (instanceContext *InstanceContext) Memory() *Memory {
 	return &instanceContext.memory
 }
 
-// Data returns the instance context data as an `interface{}`. It's
-// up to the user to cast it appropriately.
+// Data returns the instance context data as an `interface{}`. It's up to the
+// user to assert the proper type.
 func (instanceContext *InstanceContext) Data() interface{} {
-	dataPtr := *(*unsafe.Pointer)(cWasmerInstanceContextDataGet(instanceContext.context))
-
-	return *(*interface{})(dataPtr)
+	ctxDataIdx := *(*int)(cWasmerInstanceContextDataGet(instanceContext.context))
+	ctxDataMtx.RLock()
+	defer ctxDataMtx.RUnlock()
+	return ctxData[ctxDataIdx]
 }

--- a/wasmer/instance.go
+++ b/wasmer/instance.go
@@ -2,6 +2,8 @@ package wasmer
 
 import (
 	"fmt"
+	"runtime"
+	"sync"
 	"unsafe"
 )
 
@@ -72,11 +74,7 @@ type Instance struct {
 	// The exported memory of a WebAssembly instance.
 	Memory *Memory
 
-	// `contextDataC` and `contextDataGo` ensure thata`
-	// InstanceContext` data is not garbage-collected // for the
-	// lifetime of the `Instance`.
-	contextDataC  *uintptr
-	contextDataGo interface{}
+	ctxDataIdx *int
 }
 
 // NewInstance constructs a new `Instance` with no imported functions.
@@ -406,6 +404,17 @@ func (instance *Instance) HasMemory() bool {
 	return nil != instance.Memory
 }
 
+var (
+	// In order to avoid passing illegal Go pointers across the CGo FFI,
+	// store Instance Context Data in ctxData and simply pass the index
+	// through the FFI instead.
+	//
+	// See Instance.SetContextData and InstanceContext.Data.
+	ctxData        = make(map[int]interface{})
+	nextCtxDataIdx int
+	ctxDataMtx     sync.RWMutex
+)
+
 // SetContextData assigns a data that can be used by all imported functions.
 // Each imported function receives as its first argument an instance context
 // (see `InstanceContext`). An instance context can hold any kind of data,
@@ -413,16 +422,32 @@ func (instance *Instance) HasMemory() bool {
 // with reference types or pointers. It is important to understand that data is
 // global to the instance, and thus is shared by all imported functions.
 func (instance *Instance) SetContextData(data interface{}) {
-	// Cache the data with the `new(uintptr)` to prevent it to be
-	// garbage collected for the lifetime of the instance.
-	instance.contextDataGo = data
-	instance.contextDataC = new(uintptr)
-	*instance.contextDataC = uintptr(unsafe.Pointer(&instance.contextDataGo))
+	ctxDataMtx.Lock()
+	if instance.ctxDataIdx == nil {
+		instance.ctxDataIdx = new(int)
+		*instance.ctxDataIdx = nextCtxDataIdx
+		nextCtxDataIdx++
 
-	cWasmerInstanceContextDataSet(
-		instance.instance,
-		unsafe.Pointer(instance.contextDataC),
-	)
+		// When instance is GC'd, clean up its ctxData.
+		// Set the finalizer on the unexported instance.ctxDataIdx,
+		// instead of directly on the instance, to allow users of this
+		// package to set their own finalizer on the Instance for other
+		// reasons.
+		runtime.SetFinalizer(instance.ctxDataIdx, func(idx *int) {
+			// Launch a goroutine to avoid blocking other
+			// finalizers while waiting for the mutex lock.
+			go func() {
+				ctxDataMtx.Lock()
+				delete(ctxData, *idx)
+				ctxDataMtx.Unlock()
+			}()
+		})
+	}
+	ctxData[*instance.ctxDataIdx] = data
+	ctxDataMtx.Unlock()
+
+	cWasmerInstanceContextDataSet(instance.instance,
+		unsafe.Pointer(instance.ctxDataIdx))
 }
 
 // Close closes/frees an `Instance`.

